### PR TITLE
Add health check tool

### PR DESCRIPTION
**Pull request by michalzurawski**
It runs on separate port and allows checking condition of etcd and db.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -398,6 +401,18 @@ func NewServer(configFile string) (*Server, error) {
...
+	var gohanPort int
+	addressAndPort := strings.Split(server.address, ":")
+	if len(addressAndPort) != 2 {
+		panic("Incorrect gohan server address: " + server.address)
+	}
+	gohanPort, err = strconv.Atoi(addressAndPort[1])
+	if err != nil {
+		panic("Incorrect gohan server address: expected port number got " + addressAndPort[1])
+	}
+	healthCheckAddress := config.GetString("healtcheck/address", fmt.Sprintf("%s:%d", addressAndPort[0], gohanPort+1))
```
Comment (morrisson) [inline server/server.go]: comment 1: typo `healtcheck`

comment 2: How should we handle the case where healthcheck port is configured as the same value as server port?

File: healthcheck/healthcheck.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+			healthCheck.logger.Error("healthcheck error: %s", err.Error())
+			w.WriteHeader(http.StatusServiceUnavailable)
+		}
+
+	}
+	http.HandleFunc("/healthcheck", healthCheckHandler)
+	go func() {
+		for {
+			err := http.ListenAndServe(healthCheck.address, nil)
+			if err != nil {
```
Comment (p-kozlowski) [inline healthcheck/healthcheck.go]: please add a small delay between retries, I think `<-time.After(5 * time.Second)` will do

File: server/server.go [hunk 3]
```diff
@@ -398,6 +401,18 @@ func NewServer(configFile string) (*Server, error) {
 			SkipLogging: true,
 		}))
 	}
+	etcdKey := config.GetString("healthcheck/etcd_key", "/gohan")
```
Comment (p-kozlowski) [inline server/server.go]: `NewServer` is more than 200 lines long. Could we move this config parsing inside `NewHealthCheck`?

File: server/server.go [hunk 4]
```diff
@@ -398,6 +401,18 @@ func NewServer(configFile string) (*Server, error) {
...
+	addressAndPort := strings.Split(server.address, ":")
+	if len(addressAndPort) != 2 {
+		panic("Incorrect gohan server address: " + server.address)
+	}
+	gohanPort, err = strconv.Atoi(addressAndPort[1])
+	if err != nil {
+		panic("Incorrect gohan server address: expected port number got " + addressAndPort[1])
+	}
+	healthCheckAddress := config.GetString("healtcheck/address", fmt.Sprintf("%s:%d", addressAndPort[0], gohanPort+1))
+	server.Healthcheck = healthcheck.NewHealthCheck(server.db, server.sync, etcdKey, healthCheckAddress, log)
```
Comment (p-kozlowski) [inline server/server.go]: Now there's no way to disable the healthcheck. It's an unauthenticated endpoint, so it's a bit serious. I think we should add a configuration knob with disabled as default.

File: server/server.go (hunk 1 above)
Comment (michalzurawski) [reply server/server.go]: done, added error

File: healthcheck/healthcheck.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	if len(addressAndPort) != 2 {
+		return "", errors.New("Incorrect gohan server address: " + serverAddress)
+	}
+	gohanPort, err := strconv.Atoi(addressAndPort[1])
+	if err != nil {
+		return "", errors.New("Incorrect gohan server address: expected port number got " + addressAndPort[1])
+	}
+	healthCheckAddress := config.GetString(healthCheckAddressKey, fmt.Sprintf("%s:%d", addressAndPort[0], gohanPort+1))
+	if healthCheckAddress == serverAddress {
+		return "", errors.New("HealthCheck address must be different then server address " + serverAddress)
```
Comment (cierpuchaw) [inline healthcheck/healthcheck.go]: different *than* server address

File: healthcheck/healthcheck.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+func (healthCheck *HealthCheck) Run() {
+	healthCheckHandler := func(w http.ResponseWriter, req *http.Request) {
+		if err := healthCheck.isHealthy(); err == nil {
+			w.WriteHeader(http.StatusOK)
+		} else {
+			healthCheck.logger.Error("Health Check error: %s", err.Error())
+			w.WriteHeader(http.StatusServiceUnavailable)
+		}
+
```
Comment (cierpuchaw) [inline healthcheck/healthcheck.go]: Empty line between ending curly brackets.

File: healthcheck/healthcheck.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	DataStore db.DB
+	Etcd      sync.Sync
+	etcdKey   string
+	address   string
+	timeout   time.Duration
+	logger    log.Logger
+}
+
+func NewHealthCheck(db db.DB, etcd sync.Sync, serverAddress string, config *util.Config) *HealthCheck {
+	if enabled := config.GetBool(healthCheckEnabledKey, false); !enabled {
```
Comment (jkmar) [inline healthcheck/healthcheck.go]: this `enabled` seems redundant

Comment (cierpuchaw) [reply healthcheck/healthcheck.go]: ` if config.GetBool() {` should be enough.

File: server/server.go [hunk 8]
```diff
@@ -420,6 +423,9 @@ func (server *Server) SetRunning(running bool) {
 
 //Start starts GohanAPIServer
 func (server *Server) Start() (err error) {
+	if server.HealthCheck != nil {
```
Comment (p-kozlowski) [inline server/server.go]: let's make `nil` meaningful (https://goplay.space/#GjkirrWI53b)

File: healthcheck/healthcheck.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+func NewHealthCheck(db db.DB, etcd sync.Sync, serverAddress string, config *util.Config) *HealthCheck {
+	if !config.GetBool(healthCheckEnabledKey, false) {
+		return nil
+	}
+	healthCheckAddress, err := getHealthCheckAddress(serverAddress, config)
+	if err != nil {
+		panic(err)
+	}
+	timeout := time.Duration(config.GetInt(healthCheckTimeoutKey, defaultTimeoutInMs)) * time.Millisecond
```
Comment (p-kozlowski) [inline healthcheck/healthcheck.go]: actually, I won't mind if you borrow `GetDuration` from https://github.com/cloudwan/gohan/pull/715/files#diff-d891f1defb2f0e4c0d8b9494144c191aR167 ;)

Comment (michalzurawski) [reply healthcheck/healthcheck.go]: nice - stolen

**Pull request merged**